### ISSUE TITLE
Add support for interpolation to `@slice`

### ISSUE THREAD

**Issue by kdpsingh** [TidierOrg/TidierData.jl#73]
This is important for when using slice inside of for-loops.

**Comment by kdpsingh** [comment]
Right now, the use of interpolation isn't possible due to our use of `eval()` inside of `@slice()`. The reason we use `eval()` is because we can't look into the data frame to count the number of rows if the user were to use `n()`.

I think we may be able to dramatically simplify the implementation of `@slice()` to make better use of the `end` keyword.

We would return `:end` using `parse_interpolation()` whenever it detects `n()` and then we would do something like this, keeping in mind that we also need to handle negated numbers.

```julia
julia> df = DataFrame(a = repeat('a':'e', inner = 3), b = 1:15)

julia> df[[x isa UnitRange ? collect(x) : x for x in [1, 1:2, end, 3]] |> x -> vcat(x...),:]
5×2 DataFrame
 Row │ a     b     
     │ Char  Int64 
─────┼─────────────
   1 │ a         1
   2 │ a         1
   3 │ a         2
   4 │ e        15
   5 │ a         3
```

**Comment by kdpsingh** [comment]
Here's a potentially better/faster approach:

```julia
julia> df[Iterators.flatten([1, 1:2, end, 3]) |> collect,:]
5×2 DataFrame
 Row │ a     b     
     │ Char  Int64 
─────┼─────────────
   1 │ a         1
   2 │ a         1
   3 │ a         2
   4 │ e        15
   5 │ a         3
```

**Comment by drizk1** [comment]
Interesting. I was just working on adding interpolation  and n ()support to `@slice_tail` and `@slice_head`

the below all run and return the correct df even with the use of `eval` in the macro.
```
aa = 1 
@chain df begin
  @group_by(dt5)
  @slice_tail2(n = aa)
  @ungroup
end

@chain df begin
  @group_by(dt5)
  @slice_tail2(n = n()-2)
  @ungroup
end
```
`eval` code expert
```
if all(clean_indices .> 0)
          if $(esc(df)) isa GroupedDataFrame
              combine($(esc(df)); ungroup = false) do sdf
                  local n_rows_group = nrow(sdf)
                  local interpolated_indices = parse_slice_n.($exprs, n_rows_group)
                  local original_indices = [eval.(interpolated_indices)...]
                  local clean_indices = Int64[]
```

**Comment by kdpsingh** [comment]
Also, we can't rely on `parse_tidy()` for row indexing because `Between()` isn't supported for the row indexing portion.

However, if we can parse the expression to detect negation, then we *can* interpolate it like this:

```julia
julia> df[Not(collect(Iterators.flatten([1,1:2,end,3]))),:]
```

**Comment by kdpsingh** [comment]
Fixed in #75.
